src/main/java/com/microfocus/application/automation/tools/results/parser/nunit3/NUnit3ReportParserImpl.java

```diff
@@ -0,0 +1,82 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.microfocus.application.automation.tools.results.parser.nunit3;
+
+import com.microfocus.application.automation.tools.results.parser.ReportParseException;
+import com.microfocus.application.automation.tools.results.parser.ReportParser;
+import com.microfocus.application.automation.tools.results.parser.antjunit.AntJUnitReportParserImpl;
+import com.microfocus.application.automation.tools.results.service.almentities.AlmTestSet;
+import hudson.FilePath;
+import org.apache.commons.io.IOUtils;
+
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.transform.stream.StreamSource;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.util.List;
+
+public class NUnit3ReportParserImpl implements ReportParser {
+
+    private static final String TEMP_JUNIT_FILE_PREFIX = "temp-junit";
+    private static final String TEMP_JUNIT_FILE_SUFFIX = ".xml";
+    private static final String NUNIT_TO_JUNIT_XSLFILE_STR = "nunit-to-junit.xsl";
+
+    private FilePath workspace;
+
+    public NUnit3ReportParserImpl(FilePath workspace) {
+        this.workspace = workspace;
+    }
+
+    @Override
+    public List<AlmTestSet> parseTestSets(InputStream reportInputStream, String testingFramework, String testingTool)
+            throws ReportParseException {
+
+        // Use the xsl to convert the nunit3 and nunit to junit and then parse with junit logic.
+        // This can be extended to cover all kinds of result format.
+        // When new format comes, only need to provide a xsl, no need to change any code.
+
+        FileOutputStream fileOutputStream = null;
+        try {
+            TransformerFactory transformerFactory = TransformerFactory.newInstance();
+            Transformer nunitTransformer = transformerFactory.newTransformer(
+                        new StreamSource(this.getClass().getResourceAsStream(NUNIT_TO_JUNIT_XSLFILE_STR)));
+            File junitTargetFile = new File(workspace.createTempFile(TEMP_JUNIT_FILE_PREFIX, TEMP_JUNIT_FILE_SUFFIX).toURI());
+            fileOutputStream = new FileOutputStream(junitTargetFile);
+            nunitTransformer.transform(new StreamSource(reportInputStream), new StreamResult(fileOutputStream));
+
+            InputStream in = new FileInputStream(junitTargetFile);
+            return new AntJUnitReportParserImpl().parseTestSets(in, testingFramework, testingTool);
+
+        } catch (Exception e) {
```

Comment:
Try catching the specific exception that can occur instead of everyone, not everything that you will catch means that it's ReportParseException. 
Contact me regarding this, I would like to hear your opinion.

Reply:
I just follow the other report parser. And the parsers are in a list. Take turns to parse the report. If it can't it will fail. We don't care why it fail.

Reply:
Yeah, but why if it fails for a different reason that is unconnected to the parsing of the report?
Would it still be a ReportParseException?

Reply:
Thought I didn't write this, but I think the original developer was using ReportParserException just to wrap the exception up.